### PR TITLE
Fix ELB cleanup failure due to resource name end with a hyphen

### PR DESCRIPTION
**Pull request by 0xlen**
Initiate the bug fix thread to mitigate the ELB cleanup issue due to incorrect resource name. (See #7028)

### Description

The current ELB cleanup logic doesn't handle the case when the trimmed ELB hostname string is ending with hyphen (`-`). This can cause the eksctl throw the error when making DescribeLoadBalancers API calls due to incorrect resource format. The issue can be reproduced by using the testing file [1] with the following error:

```
$ eksctl version
0.154.0
$ eksctl delete cluster --region us-west-2 --name eks
2023-08-31 00:44:39 [ℹ]  deleting EKS cluster "eks"
2023-08-31 00:44:40 [ℹ]  will drain 0 unmanaged nodegroup(s) in cluster "eks"
2023-08-31 00:44:40 [ℹ]  starting parallel draining, max in-flight of 1
2023-08-31 00:44:41 [ℹ]  deleted 0 Fargate profile(s)
2023-08-31 00:44:42 [ℹ]  cleaning up AWS load balancers created by Kubernetes objects of Kind Service or Ingress
Error: cannot obtain information for ALB from Ingress default/testcase-0: cannot obtain security groups for ALB k8s-default-testcase-93d8419948-: cannot describe ELB: operation error Elastic Load Balancing v2: DescribeLoadBalancers, https response error StatusCode: 400, RequestID: ......, api error ValidationError: The load balancer name 'k8s-default-testcase-93d8419948-' cannot end with a hyphen(-)
```

By fixing the resource name issue, the expected deletion can be successfully executed [2].

[1] https://gist.github.com/0xlen/d3bab9e46a8e797760e4731817543b28
[2] https://gist.github.com/0xlen/15cd7f43fca4cc8483e70ed7fc85bcb5

### Checklist
- [X] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [X] Manually tested
- [X] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [X] Backfilled missing tests for code in same general area :tada:
- [X] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by TiberiuGC** [comment]
Hi @0xlen , thanks for opening a PR for this issue! 🚀 

Before approving I wanted to confirm that trimming the ending `-` now properly covers all possible naming formats. Are you positive about that?

File: pkg/elb/cleanup_test.go
Comment (a-hilaly) [inline pkg/elb/cleanup_test.go:46]: Tiny suggestion:
```go
			testCases := [][]struct{hostname, metadataName, expected string}{
				{
				    "bf647c9e-default-appingres-350b-1622159649.eu-central-1.elb.amazonaws.com", 
				    "",
				    "bf647c9e-default-appingres-350b",
				  },
				{
				    "internal-k8s-default-testcase-93d8419948-541000771.us-west-2.elb.amazonaws.com",
				    "",
				    "k8s-default-testcase-93d8419948",
				},
                // ...
			}

			for i, tc := range testCases {
				name := getIngressELBName(context.TODO(), tc.metadatName, []string{tc.hostname})
				Expect(name).To(Equal(tc.expected))
			}
```

File: pkg/elb/cleanup.go
Comment (a-hilaly) [inline pkg/elb/cleanup.go:217]: Do we need this if we already remove the ELB ID?

**Review by a-hilaly — COMMENTED** [review]
Thank you @0xlen ! Left few questions/suggestions below.

File: pkg/elb/cleanup.go
Comment (0xlen) [reply pkg/elb/cleanup.go]: Thanks @a-hilaly, that's a great point! I tried to revise this part and let `getIngressLoadBalancer` handle it. To have no interruption deletion of `eksctl delete` command, I think it would be better to keep the current business logic to skip the ALB resource clean up [1] if there's unexpected error when getting ELB name.

https://github.com/eksctl-io/eksctl/blob/6e8e73c13cee21b3a833f111e63ade8c03536d04/pkg/elb/cleanup.go#L220-L224

Update: https://github.com/eksctl-io/eksctl/pull/7029/commits/b638eacf9e0b8d75ae174a7a729b59d4cf2a4550#diff-bf8761431813288f0320030b76870e9d4a40dd0e80e1ca725ef262184c498b90R237-R241

File: pkg/elb/cleanup_test.go
Comment (0xlen) [reply pkg/elb/cleanup_test.go]: Thanks for the advice, this has been revised here: https://github.com/eksctl-io/eksctl/pull/7029/commits/b638eacf9e0b8d75ae174a7a729b59d4cf2a4550#diff-581a28d52a0e2589ad7027966d2b578dfe8dce701181ce35e53c2d4b06f5b302

File: pkg/elb/cleanup.go
Comment (0xlen) [reply pkg/elb/cleanup.go]: That's a good question. From my research ELB document [1] explicitly mentioned the rule when giving the name of load balancer, I think we still don't expect there's an ELB name exceeds 32 characters.

It might be better revise this part as exception case and throw the error directly, and let `getIngressLoadBalancer ` handle proper action (like skip ALB deletion right now, or interrupt the clean up process in the future).

Update: https://github.com/eksctl-io/eksctl/pull/7029/commits/4989bf86801145276785f5691c310e808f33cf12#diff-bf8761431813288f0320030b76870e9d4a40dd0e80e1ca725ef262184c498b90R214-R215

[1] https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_CreateLoadBalancer.html#API_CreateLoadBalancer_RequestParameters

File: pkg/elb/cleanup.go
Comment (0xlen) [reply pkg/elb/cleanup.go]: Thanks for the suggestion! But I am not sure if this update is what you are looking for? Or does it mean the debug message of showing ALB resource name should be moved to `getIngressLoadBalancer`?

- https://github.com/eksctl-io/eksctl/pull/7029/commits/4989bf86801145276785f5691c310e808f33cf12#diff-bf8761431813288f0320030b76870e9d4a40dd0e80e1ca725ef262184c498b90R217

I did not fully understand what could be the difference as it seems like not all logger in the package is applying the defer directive, I assume it might be some performance improvement, or, it helps the readability (which I think would be what you are mentioning). Any details would be appreciated!

Comment (0xlen) [reply pkg/elb/cleanup.go]: Moved the logger to `getIngressLoadBalancer`: https://github.com/eksctl-io/eksctl/compare/4989bf86801145276785f5691c310e808f33cf12..b73c890febc5feec8b4ae554067260aecfe73a91

File: pkg/elb/cleanup.go
Comment (a-hilaly) [inline pkg/elb/cleanup.go:202]: We can remove `ctx context.Context` from the params

File: pkg/elb/cleanup.go
Comment (a-hilaly) [inline pkg/elb/cleanup.go:240]: Maybe we should error if the `parsed name exceeds maximum of 32 characters` is returned? thoughts @cPu1 ?

**Review by a-hilaly — APPROVED** [review]
Great stuff @0xlen , thank you! I have one nit/question below

File: pkg/elb/cleanup.go
Comment (0xlen) [reply pkg/elb/cleanup.go]: Agree, I removed the unused `ctx` parameter.

File: pkg/elb/cleanup.go
Comment (0xlen) [reply pkg/elb/cleanup.go:240]: I was referring original general text as considering there can have `cannot get the hostname` or `parsed name exceeds maximum of 32 characters`. Would be interested to know what could be the proper error message and I am happy to revise that.

**Comment by a-hilaly** [comment]
@0xlen Thank you for iterating on this! LGTM. I'll let the eksctl experts have a say on this before we merge it.

**Review by TiberiuGC — APPROVED** [review]
LGTM 👍

**Pull request merged**
